broadinstitute/single_cell_portal_core
# Instant gene expression scatter plots: prototype (SCP-4583)

This adds progressive loading.  Gene expression scatter plots now display instantly, even for extremely many cells.

Previously, gene expression scatter plots had to be fully interactive in order to be visually displayed.  [Time-to-view can be decoupled from time-to-interactive](https://github.com/broadinstitute/single_cell_portal_core/pull/1571).  In theory, that massively improves render times.  This work implements that in practice.

Here's how it looks:

https://user-images.githubusercontent.com/1334561/185487425-7a7d450b-394c-4987-a711-7fa7cb9f2c01.mov

The plot displayed 0.5 seconds after searching is a static image.  It's visible, but not interactive.  The nearly identical plot with tooltips shown ~5 seconds after searching is the interactive version. 

Graphical scaling and transforming will be refined in SCP-4600.  That'll fix the slight jank / point-mass shift upon transition.

To test:
* Run `rake db:migrate`
* Find the accession for your local "Human milk -- differential expression" study
  * If you don't have that study, you can add it via Terminal like so: `echo "SyntheticStudyPopulator.populate('human_milk_de_49k_cell')" | bundle exec rails console -e development`
* Enable "progressive_loading" feature flag for the study
* In the study, search "A1BG-AS1"
* Quickly hover over points in the gene expression scatter plot, note no tooltips
* Confirm slight change in plot, and appearance of tooltips on hover

This satisfies SCP-4583.
